consensus/src/metrics_safety_rules.rs

```diff
@@ -35,16 +35,30 @@ impl MetricsSafetyRules {
     pub fn perform_initialize(&mut self) -> Result<(), Error> {
         let consensus_state = self.consensus_state()?;
         let sr_waypoint = consensus_state.waypoint();
-        let proofs = self
-            .storage
-            .retrieve_epoch_change_proof(sr_waypoint.version())
-            .map_err(|e| {
-                Error::InternalError(format!(
-                    "Unable to retrieve Waypoint state from storage, encountered Error:{}",
-                    e
-                ))
-            })?;
-        self.initialize(&proofs)
+        loop {
+            let proofs = self
+                .storage
+                .retrieve_epoch_change_proof(sr_waypoint.version())
```

Comment:
we should use the updated version to retrieve the proof, otherwise this would exceed the storage limit and won't paginate

Reply:
@zekun000, good catch! So you mean move lines 36 and 37 into this loop, right? i.e., refetch the latest waypoint each time?

Reply:
or use the curr_version returned carried in the error to avoid another call to safety rules

Reply:
Fixed by using the updated version curr_version to retrieve proof at the beginning of each loop.

Reply:
FYI: @khiemngo, you'll want to squash all of your commits into a single commit so that the git history is cleaner. 😄 Reviewers will always review the final product, so don't worry about making separate commits.

Reply:
@JoshLind Ahhh this is great, thanks!! I've just squashed the recent commits into one as you suggested.